src/fairchem/applications/fastcsp/core/utils/structure.py

```diff
@@ -0,0 +1,276 @@
+        vol_bin = round(volume / vol_bin_size**3) * vol_bin_size**3
+        hash_components.append(f"{vol_bin:.1f}")
+
+    # Combine all components into single hash string
+    return "_".join(hash_components)
+
+
+def check_no_changes_in_covalent_matrix(
+    initial_atoms: Atoms, final_atoms: Atoms
+) -> bool:
+    """
+    Validate that covalent bonding network is preserved during structure relaxation.
+
+    Compares the covalent bonding adjacency matrices before and after ML-based
+    relaxation to detect unwanted chemical reconstructions. This validation ensures
+    that the relaxation process only optimizes geometry without breaking or forming
+    chemical bonds, which would indicate problematic initial structures or
+    relaxation failures.
+
+    Args:
+        initial_atoms: Original structure before relaxation
+        final_atoms: Structure after ML-based relaxation
+
+    Returns:
+        True if bonding network is preserved, False otherwise
+        Returns False if either structure is None (error handling)
+
+    Algorithm:
+        1. Convert ASE Atoms to pymatgen Structures for analysis
+        2. Use JmolNN to identify covalent neighbors in both structures
+        3. Build adjacency matrices representing bonding networks
+        4. Compare matrices for exact equality
+
+    Validation Purpose:
+        - Detect atom overlaps that lead to artificial bonding
+        - Identify relaxation artifacts that break molecular integrity
+        - Filter out reconstructions that change chemical connectivity
+    """
+    # Handle error cases where structures couldn't be processed
+    if initial_atoms is None or final_atoms is None:
+        return False
+
+    # Convert ASE Atoms to pymatgen Structures for neighbor analysis
+    initial_structure = AseAtomsAdaptor.get_structure(initial_atoms)
+    final_structure = AseAtomsAdaptor.get_structure(final_atoms)
+
+    # Build adjacency matrix for initial structure using Jmol bonding radii
+    initial_nn_info = JmolNN().get_all_nn_info(initial_structure)
+    initial_nn_matrix = np.zeros((len(initial_nn_info), len(initial_nn_info)))
+    for i in range(len(initial_nn_info)):
+        for j in range(len(initial_nn_info[i])):
+            # Mark bonded pairs in adjacency matrix
+            initial_nn_matrix[i, initial_nn_info[i][j]["site_index"]] = 1
+
+    # Build adjacency matrix for final (relaxed) structure
+    final_nn_info = JmolNN().get_all_nn_info(final_structure)
+    final_nn_matrix = np.zeros((len(final_nn_info), len(final_nn_info)))
+    for i in range(len(final_nn_info)):
+        for j in range(len(final_nn_info[i])):
+            # Mark bonded pairs in adjacency matrix
+            final_nn_matrix[i, final_nn_info[i][j]["site_index"]] = 1
+
+    # Check that both bonding networks are identical
+    # Any difference indicates bond formation/breaking during relaxation
+    return np.array_equal(initial_nn_matrix, final_nn_matrix)
+
+
+def check_no_changes_in_Z(
```

Comment:
We can probably remove this because `check_no_changes_in_covalent_matrix` should be a stronger check.